SpatialGDK/Source/Private/Interop/SnapshotManager.cpp

```diff
@@ -0,0 +1,191 @@
+
+	Worker_EntityQuery WorldQuery{};
+	WorldQuery.constraint = WorldConstraint;
+	WorldQuery.result_type = WORKER_RESULT_TYPE_SNAPSHOT;
+
+	Worker_RequestId RequestID;
+	RequestID = NetDriver->Connection->SendEntityQueryRequest(&WorldQuery);
+
+	EntityQueryDelegate WorldQueryDelegate;
+	WorldQueryDelegate.BindLambda([this, Delegate](Worker_EntityQueryResponseOp& Op)
+	{
+		if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+		{
+			UE_LOG(LogSnapshotManager, Error, TEXT("SnapshotManager WorldWipe - World entity query failed: %s"), UTF8_TO_TCHAR(Op.message));
+		}
+
+		if (Op.result_count == 0)
+		{
+			UE_LOG(LogSnapshotManager, Error, TEXT("SnapshotManager WorldWipe - No entities found in world entity query"));
+		}
+
+		if (Op.result_count >= 1)
+		{
+			// Send deletion requests for all entities found in the world entity query.
+			DeleteEntities(Op);
+
+			// Also make sure that we kill the GSM.
+			NetDriver->Connection->SendDeleteEntityRequest(GlobalStateManager->GlobalStateManagerEntityId);
+		}
+
+		Delegate.ExecuteIfBound();
+	});
+
+	Receiver->AddEntityQueryDelegate(RequestID, WorldQueryDelegate);
+}
+
+void USnapshotManager::DeleteEntities(const Worker_EntityQueryResponseOp& Op)
+{
+	UE_LOG(LogSnapshotManager, Warning, TEXT("Deleting %u entities."), Op.result_count);
+
+	for (uint32_t i = 0; i < Op.result_count; i++)
+	{
+		UE_LOG(LogSnapshotManager, Log, TEXT("Sending delete request for: %i"), Op.results[i].entity_id);
+		NetDriver->Connection->SendDeleteEntityRequest(Op.results[i].entity_id);
+	}
+}
+
+// GetSnapshotPath will take a snapshot (without the .snapshot extension) name and convert it to a relative path in the Game/Content folder.
+FString GetSnapshotPath(const FString& SnapshotName)
+{
+	FString SnapshotsDirectory = FPaths::ProjectContentDir().Append("Spatial/Snapshots/");
+	return SnapshotsDirectory + SnapshotName + ".snapshot";
+}
+
+// LoadSnapshot will take a snapshot name which should be on disk and attempt to read and spawn all of the entities in that snapshot.
+// This should only be called from the worker which has authority over the GSM.
+void USnapshotManager::LoadSnapshot(const FString& SnapshotName)
+{
+	FString SnapshotPath = GetSnapshotPath(SnapshotName);
+
+	UE_LOG(LogSnapshotManager, Warning, TEXT("Loading snapshot: '%s'"), *SnapshotPath);
+
+	Worker_ComponentVtable DefaultVtable{};
+	Worker_SnapshotParameters Parameters{};
+	Parameters.default_component_vtable = &DefaultVtable;
+
+	Worker_SnapshotInputStream* Snapshot = Worker_SnapshotInputStream_Create(TCHAR_TO_UTF8(*SnapshotPath), &Parameters);
+
+	FString Error = Worker_SnapshotInputStream_GetError(Snapshot);
+	if (!Error.IsEmpty())
+	{
+		UE_LOG(LogSnapshotManager, Error, TEXT("Error when attempting to read snapshot '%s': %s"), *SnapshotPath, *Error);
+		return;
+	}
+
+	TArray<TArray<Worker_ComponentData>> EntitiesToSpawn;
+
+	// Get all of the entities from the snapshot.
+	while (Snapshot && Worker_SnapshotInputStream_HasNext(Snapshot) > 0)
```

Comment:
If Snapshot is null, wouldn't the above GetError catch it?

Reply:
Actually here we are checking the error of `HasNext`, the API expects you to check the error after all calls.